protocol-units/da/m1/light-node/src/v1/sequencer.rs

```diff
@@ -88,64 +97,77 @@ impl LightNodeV1 {
 
 	}
 
-	/// Attempts to submit a batch of blobs according to the chunk size
-	async fn submit_and_resize(&self, block_blobs: &mut Vec<celestia_types::Blob>, chunk_size : usize) -> Result<usize, anyhow::Error> {
-	
+	async fn submit_blocks(&self, blocks : &Vec<Block>) -> Result<(), anyhow::Error> {
 
-		// iter in chunks of 4
-		while !block_blobs.is_empty() {
-			let mut chunk: Vec<celestia_types::Blob> = Vec::with_capacity(chunk_size);
-			
-			for _ in 0..chunk_size {
-				if let Some(blob) = block_blobs.pop() {
-					chunk.push(blob);
-				} else {
-					break;
-				}
-			}
-			
-			match self.pass_through.submit_celestia_blobs(&chunk).await {
-				Ok(_) => {
-					info!(
-						target: "movement_timing",
-						batch_size = chunk.len(),
-						"submitted_blob_batch",
-					);
-				}
-				Err(e) => {
-					info!("failed to submit blocks: {:?}", e);
-					block_blobs.extend(chunk.into_iter());
-					return Ok(chunk_size - 1);
-				}
-			}
+		let uid = LOGGING_UID.load(Ordering::SeqCst);
+		info!(
+			target: "movement_timing",
+			batch_size = blocks.len(),
+			uid = uid, 
+			"inner_submitting_block_batch"
+		);
+		let mut block_blobs = Vec::new();
+		for block in blocks {
+			let block_blob = self.pass_through.create_new_celestia_blob(
+					bcs::to_bytes(block)
+					.map_err(|e| anyhow::anyhow!("Failed to serialize block: {}", e))?,
+			)?;
+			block_blobs.push(block_blob);
 		}
-	
+		for block in blocks {
+			info!(target: "movement_timing", block_id = %block.id(), "submitting_block");
+		}
+		self.pass_through.submit_celestia_blobs(&block_blobs).await?;
+		for block in blocks {
+			info!(target: "movement_timing", block_id = %block.id(), "submitted_block");
+		}
+		Ok(())
+	}
 
-		Ok(chunk_size)
+	pub async fn submit_with_heuristic(&self, blocks : Vec<Block>) -> Result<(), anyhow::Error> {
 
-	}
+		let mut heuristic : GroupingHeuristicStack<Block> = GroupingHeuristicStack::new(vec![
+			DropSuccess::boxed(),
+			ToApply::boxed(),
+			SkipFor::boxed(1, Splitting::boxed(2)),
+			FirstFitBinpacking::boxed(1_700_000)
+		]);
 
-	/// Submits block blobs in chunks, beginning with a chunk size of 4
-	async fn submit_block_blobs(&self, block_blobs: &mut Vec<celestia_types::Blob>) -> Result<(), anyhow::Error> {
-		if block_blobs.len() > 0 {
+		let start_distribution = GroupingOutcome::new_apply_distribution(blocks);
+		let block_group_results = heuristic.run_async_sequential_with_metadata(
+			start_distribution,
+			|index, grouping, mut flag| async move {
 
-			let mut chunk_size = 4;
-			loop {
+				if index == 0 {
+					flag = false;
+				}
```

Comment:
This seems like maybe `run_async_sequential_with_metadata` needs to pass the initial metadata value, like `Iterator::fold`. Passing the index around is not strictly necessary in the general case because you can pass an `FnMut` closure that can have synchronous code for the loop iteration, like this:

```rust
move |grouping, mut flag| {
    // logic about the index, maybe copy it to a capture var for the async move closure

    let fut = async move {
        // ...
    };

    index += 1;
    fut
}
```

Reply:
I was thinking about the same thing. Really, the metadata value performs a more general version of what the index provides, i.e., tracking and updating non-grouping data. I'll drop the index.

Reply:
Actually, I changed my mind back. `index` does provide information that would be hard to compute otherwise. For example, `index == 0` indicates the start of a new distribution being applied and would be commonly used. It probably even makes sense to provide the distribution length to this callback in the future.